plugin/code_actions.py

```diff
@@ -293,65 +253,60 @@ def run(
-
-    def handle_responses_async(self, responses: CodeActionsByConfigName, run_first: bool = False) -> None:
-        self.commands_by_config = responses
-        self.commands = self.combine_commands()
-        if len(self.commands) == 1 and run_first:
-            self.handle_select(0)
-        else:
-            self.show_code_actions()
-
-    def combine_commands(self) -> 'List[Tuple[str, CodeActionOrCommand]]':
-        results = []
-        for config, commands in self.commands_by_config.items():
-            for command in commands:
-                results.append((config, command))
-        return results
-
-    def show_code_actions(self) -> None:
-        if len(self.commands) > 0:
-            window = self.view.window()
-            if window:
-                items, selected_index = format_code_actions_for_quick_panel([command[1] for command in self.commands])
-                window.show_quick_panel(
-                    items,
-                    self.handle_select,
-                    selected_index=selected_index,
-                    placeholder="Code action")
+        if code_actions_by_config:
+            self._handle_code_actions(code_actions_by_config, run_first=True)
+            return
+        self._run_async(only_kinds)
+
+    def _run_async(self, only_kinds: Optional[List[CodeActionKind]] = None) -> None:
+        view = self.view
+        region = first_selection_region(view)
+        if region is None:
+            return
+        listener = windows.listener_for_view(view)
+        if not listener:
+            return
+        session_buffer_diagnostics, covering = listener.diagnostics_intersecting_async(region)
+        actions_manager \
+            .request_for_region_async(view, covering, session_buffer_diagnostics, only_kinds, manual=True) \
+            .then(lambda actions: sublime.set_timeout(lambda: self._handle_code_actions(actions)))
+
+    def _handle_code_actions(self, response: List[CodeActionsByConfigName], run_first: bool = False) -> None:
+        # Flatten response to a list of (config_name, code_action) tuples.
+        actions = []  # type: List[Tuple[ConfigName, CodeActionOrCommand]]
+        for config_name, session_actions in response:
+            actions.extend([(config_name, action) for action in session_actions])
+        if len(actions) == 1 and run_first:
+            self._handle_select(0, actions)
         else:
-            self.view.show_popup('No actions available', sublime.HIDE_ON_MOUSE_MOVE_AWAY)
+            self._show_code_actions(actions)
 
-    def handle_select(self, index: int) -> None:
-        if index > -1:
+    def _show_code_actions(self, actions: List[Tuple[ConfigName, CodeActionOrCommand]]) -> None:
+        window = self.view.window()
+        if not window:
+            return
+        items, selected_index = format_code_actions_for_quick_panel(actions)
+        window.show_quick_panel(
```

Comment:
`show_quick_panel` is smart enough to not do anything if `items` is empty, but it still seems wasteful to do that API call in this case. Also there needs to be some kind of feedback if no code actions are available, otherwise it's bad UX if user selects the menu entry and nothing happens (I like that the strange popup is gone now, though). A status bar message seems appropriate to me:

```python
if items:
    window.show_quick_panel(...)
else:
    window.status_message("No actions available")
```